File: packages/less/src/less/tree/detached-ruleset.test.js
import { describe, it, expect, vi } from 'vitest';
import DetachedRuleset from './detached-ruleset';
import Node from './node';
import contexts from '../contexts';

describe('DetachedRuleset', () => {
    // Mock dependencies
    const mockRuleset = Object.assign(new Node(), {
        eval: vi.fn()
    });

    const mockContext = {
        frames: ['frame1', 'frame2']
    };

    const mockVisitor = {
        visit: vi.fn()
    };

    beforeEach(() => {
        vi.clearAllMocks();
    });

    describe('constructor', () => {
        it('should create a DetachedRuleset instance with the provided ruleset', () => {
            const detachedRuleset = new DetachedRuleset(mockRuleset);
            expect(detachedRuleset.ruleset).toBe(mockRuleset);
            expect(detachedRuleset.frames).toBeUndefined();
        });

        it('should create a DetachedRuleset instance with the provided ruleset and frames', () => {
            const frames = ['frame1', 'frame2'];
            const detachedRuleset = new DetachedRuleset(mockRuleset, frames);
            expect(detachedRuleset.ruleset).toBe(mockRuleset);
            expect(detachedRuleset.frames).toBe(frames);
        });

        it('should set the parent of the ruleset to itself', () => {
            const detachedRuleset = new DetachedRuleset(mockRuleset);
            expect(mockRuleset.parent).toBe(detachedRuleset);
        });

        it('should handle null ruleset', () => {
            const detachedRuleset = new DetachedRuleset(null);
            expect(detachedRuleset.ruleset).toBeNull();
            expect(detachedRuleset.frames).toBeUndefined();
        });

        it('should handle undefined ruleset', () => {
            const detachedRuleset = new DetachedRuleset(undefined);
            expect(detachedRuleset.ruleset).toBeUndefined();
            expect(detachedRuleset.frames).toBeUndefined();
        });

        it('should handle null frames', () => {
            const detachedRuleset = new DetachedRuleset(mockRuleset, null);
            expect(detachedRuleset.frames).toBeNull();
        });

        it('should handle undefined frames', () => {
            const detachedRuleset = new DetachedRuleset(mockRuleset, undefined);
            expect(detachedRuleset.frames).toBeUndefined();
        });

        it('should handle setParent failure', () => {
            const originalSetParent = Node.prototype.setParent;
            Node.prototype.setParent = vi.fn().mockImplementation(() => {
                throw new Error('setParent failed');
            });

            expect(() => new DetachedRuleset(mockRuleset)).toThrow(
                'setParent failed'
            );

            Node.prototype.setParent = originalSetParent;
        });
    });

    describe('eval', () => {
        it('should return a new DetachedRuleset with the same ruleset when no frames provided', () => {
            const detachedRuleset = new DetachedRuleset(mockRuleset);
            const result = detachedRuleset.eval(mockContext);

            expect(result).toBeInstanceOf(DetachedRuleset);
            expect(result.ruleset).toBe(mockRuleset);
            expect(result.frames).toEqual(mockContext.frames);
        });

        it('should return a new DetachedRuleset with the same ruleset and frames when frames provided', () => {
            const frames = ['frame1', 'frame2'];
            const detachedRuleset = new DetachedRuleset(mockRuleset, frames);
            const result = detachedRuleset.eval(mockContext);

            expect(result).toBeInstanceOf(DetachedRuleset);
            expect(result.ruleset).toBe(mockRuleset);
            expect(result.frames).toBe(frames);
        });

        it('should handle undefined context frames', () => {
            const detachedRuleset = new DetachedRuleset(mockRuleset);
            const contextWithoutFrames = { frames: [] };
            const result = detachedRuleset.eval(contextWithoutFrames);

            expect(result).toBeInstanceOf(DetachedRuleset);
            expect(result.ruleset).toBe(mockRuleset);
            expect(result.frames).toEqual([]);
        });

        it('should handle null context', () => {
            const detachedRuleset = new DetachedRuleset(mockRuleset);
            expect(() => detachedRuleset.eval(null)).toThrow(
                "Cannot read properties of null (reading 'frames')"
            );
        });

        it('should handle undefined context', () => {
            const detachedRuleset = new DetachedRuleset(mockRuleset);
            expect(() => detachedRuleset.eval(undefined)).toThrow(
                "Cannot read properties of undefined (reading 'frames')"
            );
        });

        it('should handle context with undefined frames', () => {
            const detachedRuleset = new DetachedRuleset(mockRuleset);
            const contextWithUndefinedFrames = {};
            expect(() =>
                detachedRuleset.eval(contextWithUndefinedFrames)
            ).toThrow("Cannot read properties of undefined (reading 'length')");
        });

        it('should handle empty frames array in context', () => {
            const detachedRuleset = new DetachedRuleset(mockRuleset);
            const contextWithEmptyFrames = { frames: [] };
            const result = detachedRuleset.eval(contextWithEmptyFrames);

            expect(result).toBeInstanceOf(DetachedRuleset);
            expect(result.ruleset).toBe(mockRuleset);
            expect(result.frames).toEqual([]);
        });
    });

    describe('callEval', () => {
        it('should evaluate the ruleset with concatenated frames when frames exist', () => {
            const frames = ['frame1', 'frame2'];
            const detachedRuleset = new DetachedRuleset(mockRuleset, frames);
            const expectedContext = new contexts.Eval(
                mockContext,
                frames.concat(mockContext.frames)
            );

            detachedRuleset.callEval(mockContext);
            expect(mockRuleset.eval).toHaveBeenCalledWith(expectedContext);
        });

        it('should evaluate the ruleset with original context when no frames exist', () => {
            const detachedRuleset = new DetachedRuleset(mockRuleset);
            detachedRuleset.callEval(mockContext);
            expect(mockRuleset.eval).toHaveBeenCalledWith(mockContext);
        });

        it('should handle undefined context', () => {
            const detachedRuleset = new DetachedRuleset(mockRuleset);
            const result = detachedRuleset.callEval(undefined);
            expect(mockRuleset.eval).toHaveBeenCalledWith(undefined);
            expect(result).toBeUndefined();
        });

        it('should handle null context', () => {
            const detachedRuleset = new DetachedRuleset(mockRuleset);
            const result = detachedRuleset.callEval(null);
            expect(mockRuleset.eval).toHaveBeenCalledWith(null);
            expect(result).toBeUndefined();
        });

        it('should handle ruleset eval throwing an error', () => {
            const errorRuleset = Object.assign(new Node(), {
                eval: vi.fn().mockImplementation(() => {
                    throw new Error('Eval failed');
                })
            });
            const detachedRuleset = new DetachedRuleset(errorRuleset);

            expect(() => detachedRuleset.callEval(mockContext)).toThrow(
                'Eval failed'
            );
        });

        it('should handle ruleset eval returning null', () => {
            const nullRuleset = Object.assign(new Node(), {
                eval: vi.fn().mockReturnValue(null)
            });
            const detachedRuleset = new DetachedRuleset(nullRuleset);

            const result = detachedRuleset.callEval(mockContext);
            expect(result).toBeNull();
        });

        it('should handle ruleset eval returning undefined', () => {
            const undefinedRuleset = Object.assign(new Node(), {
                eval: vi.fn().mockReturnValue(undefined)
            });
            const detachedRuleset = new DetachedRuleset(undefinedRuleset);

            const result = detachedRuleset.callEval(mockContext);
            expect(result).toBeUndefined();
        });

        it('should handle context with undefined frames', () => {
            const detachedRuleset = new DetachedRuleset(mockRuleset);
            const contextWithUndefinedFrames = {};
            detachedRuleset.callEval(contextWithUndefinedFrames);
            expect(mockRuleset.eval).toHaveBeenCalledWith(
                contextWithUndefinedFrames
            );
        });

        it('should handle context with null frames', () => {
            const detachedRuleset = new DetachedRuleset(mockRuleset);
            const contextWithNullFrames = { frames: null };
            detachedRuleset.callEval(contextWithNullFrames);
            expect(mockRuleset.eval).toHaveBeenCalledWith(
                contextWithNullFrames
            );
        });

        it('should handle contexts.Eval constructor failure', () => {
            const originalEval = contexts.Eval;
            contexts.Eval = vi.fn().mockImplementation(() => {
                throw new Error('Eval constructor failed');
            });

            const frames = ['frame1', 'frame2'];
            const detachedRuleset = new DetachedRuleset(mockRuleset, frames);
            expect(() => detachedRuleset.callEval(mockContext)).toThrow(
                'Eval constructor failed'
            );

            contexts.Eval = originalEval;
        });

        it('should handle frames.concat failure', () => {
            const frames = {
                concat: vi.fn().mockImplementation(() => {
                    throw new Error('concat failed');
                })
            };
            const detachedRuleset = new DetachedRuleset(mockRuleset, frames);
            expect(() => detachedRuleset.callEval(mockContext)).toThrow(
                'concat failed'
            );
        });

        it('should handle ruleset eval returning a non-null value', () => {
            const valueRuleset = Object.assign(new Node(), {
                eval: vi.fn().mockReturnValue('test value')
            });
            const detachedRuleset = new DetachedRuleset(valueRuleset);

            const result = detachedRuleset.callEval(mockContext);
            expect(result).toBe('test value');
        });
    });

    describe('accept', () => {
        it('should visit the ruleset with the provided visitor', () => {
            const detachedRuleset = new DetachedRuleset(mockRuleset);
            detachedRuleset.accept(mockVisitor);
            expect(mockVisitor.visit).toHaveBeenCalledWith(mockRuleset);
        });

        it('should handle null visitor', () => {
            const detachedRuleset = new DetachedRuleset(mockRuleset);
            expect(() => detachedRuleset.accept(null)).toThrow();
        });

        it('should handle undefined visitor', () => {
            const detachedRuleset = new DetachedRuleset(mockRuleset);
            expect(() => detachedRuleset.accept(undefined)).toThrow();
        });

        it('should handle visitor returning a different ruleset', () => {
            const detachedRuleset = new DetachedRuleset(mockRuleset);
            const newRuleset = Object.assign(new Node(), {
                eval: vi.fn()
            });
            const visitor = {
                visit: vi.fn().mockReturnValue(newRuleset)
            };
            detachedRuleset.accept(visitor);
            expect(detachedRuleset.ruleset).toBe(newRuleset);
        });
    });

    describe('inheritance', () => {
        it('should inherit from Node', () => {
            const detachedRuleset = new DetachedRuleset(mockRuleset);
            expect(detachedRuleset).toBeInstanceOf(Node);
        });

        it('should have the correct type property', () => {
            const detachedRuleset = new DetachedRuleset(mockRuleset);
            expect(detachedRuleset.type).toBe('DetachedRuleset');
        });

        it('should have evalFirst set to true', () => {
            const detachedRuleset = new DetachedRuleset(mockRuleset);
            expect(detachedRuleset.evalFirst).toBe(true);
        });
    });
});

File: packages/less/src/less/tree/detached-ruleset_test.go
package tree

import (
	"reflect" // Import reflect for deep equality checks
	"testing"
)

// Helper function for comparing slices (since Go doesn't directly compare slices for equality)
func slicesEqual(a, b []interface{}) bool {
	return reflect.DeepEqual(a, b)
}

// mockEvaluableNode is used for testing CallEval
// It embeds Node and provides an Eval method with the expected signature.
type mockEvaluableNode struct {
	*Node
	EvalFunc func(ctx interface{}) interface{} // Optional custom Eval behavior
}

// Eval implements the interface expected by CallEval
func (m *mockEvaluableNode) Eval(ctx interface{}) interface{} {
	if m.EvalFunc != nil {
		return m.EvalFunc(ctx)
	}
	// Default behavior: return the node itself
	return m
}

// newMockEvaluableNode creates a new mock node for testing
func newMockEvaluableNode() *mockEvaluableNode {
	return &mockEvaluableNode{Node: NewNode()}
}

func TestDetachedRuleset(t *testing.T) {
	// Mock dependencies
	mockRuleset := newMockEvaluableNode() // Use the new mock type
	mockContext := map[string]interface{}{
		"frames": []interface{}{"frame1", "frame2"},
	}

	t.Run("constructor", func(t *testing.T) {
		t.Run("should create a DetachedRuleset instance with the provided ruleset", func(t *testing.T) {
			detachedRuleset := NewDetachedRuleset(mockRuleset, nil)
			if detachedRuleset.Ruleset != mockRuleset {
				t.Errorf("Expected Ruleset to be %v, got %v", mockRuleset, detachedRuleset.Ruleset)
			}
			if detachedRuleset.Frames != nil {
				t.Errorf("Expected Frames to be nil, got %v", detachedRuleset.Frames)
			}
		})

		t.Run("should create a DetachedRuleset instance with the provided ruleset and frames", func(t *testing.T) {
			frames := []interface{}{"frame1", "frame2"}
			detachedRuleset := NewDetachedRuleset(mockRuleset, frames)
			if detachedRuleset.Ruleset != mockRuleset {
				t.Errorf("Expected Ruleset to be %v, got %v", mockRuleset, detachedRuleset.Ruleset)
			}
			if !slicesEqual(detachedRuleset.Frames, frames) {
				t.Errorf("Expected Frames to be %v, got %v", frames, detachedRuleset.Frames)
			}
		})

		t.Run("should set the parent of the ruleset to itself", func(t *testing.T) {
			detachedRuleset := NewDetachedRuleset(mockRuleset, nil)
			// Check against the embedded Node of the mock
			if mockRuleset.Node.Parent != detachedRuleset.Node {
				t.Errorf("Expected mockRuleset.Node.Parent to be %v, got %v", detachedRuleset.Node, mockRuleset.Node.Parent)
			}
		})

		t.Run("should handle null ruleset", func(t *testing.T) {
			detachedRuleset := NewDetachedRuleset(nil, nil)
			if detachedRuleset.Ruleset != nil {
				t.Errorf("Expected Ruleset to be nil, got %v", detachedRuleset.Ruleset)
			}
			if detachedRuleset.Frames != nil {
				t.Errorf("Expected Frames to be nil, got %v", detachedRuleset.Frames)
			}
		})

		t.Run("should handle undefined ruleset (same as null in Go)", func(t *testing.T) {
			detachedRuleset := NewDetachedRuleset(nil, nil)
			if detachedRuleset.Ruleset != nil {
				t.Errorf("Expected Ruleset to be nil, got %v", detachedRuleset.Ruleset)
			}
			if detachedRuleset.Frames != nil {
				t.Errorf("Expected Frames to be nil, got %v", detachedRuleset.Frames)
			}
		})

		t.Run("should handle null frames", func(t *testing.T) {
			detachedRuleset := NewDetachedRuleset(mockRuleset, nil)
			if detachedRuleset.Frames != nil {
				t.Errorf("Expected Frames to be nil, got %v", detachedRuleset.Frames)
			}
		})

		t.Run("should handle undefined frames (same as null in Go)", func(t *testing.T) {
			detachedRuleset := NewDetachedRuleset(mockRuleset, nil)
			if detachedRuleset.Frames != nil {
				t.Errorf("Expected Frames to be nil, got %v", detachedRuleset.Frames)
			}
		})
	})

	t.Run("eval", func(t *testing.T) {
		t.Run("should return a new DetachedRuleset with the same ruleset when no frames provided", func(t *testing.T) {
			detachedRuleset := NewDetachedRuleset(mockRuleset, nil)
			result := detachedRuleset.Eval(mockContext)

			resultDR, ok := result.(*DetachedRuleset)
			if !ok {
				t.Fatalf("Expected result to be of type *DetachedRuleset, got %T", result)
			}
			if resultDR.Ruleset != mockRuleset {
				t.Errorf("Expected Ruleset to be %v, got %v", mockRuleset, resultDR.Ruleset)
			}
			expectedFrames := mockContext["frames"].([]interface{})
			if !slicesEqual(resultDR.Frames, expectedFrames) {
				t.Errorf("Expected Frames to be %v, got %v", expectedFrames, resultDR.Frames)
			}
		})

		t.Run("should return a new DetachedRuleset with the same ruleset and frames when frames provided", func(t *testing.T) {
			frames := []interface{}{"frame1", "frame2"}
			detachedRuleset := NewDetachedRuleset(mockRuleset, frames)
			result := detachedRuleset.Eval(mockContext)

			resultDR, ok := result.(*DetachedRuleset)
			if !ok {
				t.Fatalf("Expected result to be of type *DetachedRuleset, got %T", result)
			}
			if resultDR.Ruleset != mockRuleset {
				t.Errorf("Expected Ruleset to be %v, got %v", mockRuleset, resultDR.Ruleset)
			}
			if !slicesEqual(resultDR.Frames, frames) {
				t.Errorf("Expected Frames to be %v, got %v", frames, resultDR.Frames)
			}
		})

		t.Run("should handle undefined context frames (empty slice)", func(t *testing.T) {
			detachedRuleset := NewDetachedRuleset(mockRuleset, nil)
			contextWithoutFrames := map[string]interface{}{"frames": []interface{}{}}
			result := detachedRuleset.Eval(contextWithoutFrames)

			resultDR, ok := result.(*DetachedRuleset)
			if !ok {
				t.Fatalf("Expected result to be of type *DetachedRuleset, got %T", result)
			}
			if resultDR.Ruleset != mockRuleset {
				t.Errorf("Expected Ruleset to be %v, got %v", mockRuleset, resultDR.Ruleset)
			}
			if len(resultDR.Frames) != 0 {
				t.Errorf("Expected Frames to be empty, got %v", resultDR.Frames)
			}
		})

		t.Run("should handle null context", func(t *testing.T) {
			detachedRuleset := NewDetachedRuleset(mockRuleset, nil)
			result := detachedRuleset.Eval(nil)

			resultDR, ok := result.(*DetachedRuleset)
			if !ok {
				t.Fatalf("Expected result to be of type *DetachedRuleset, got %T", result)
			}
			if resultDR.Ruleset != mockRuleset {
				t.Errorf("Expected Ruleset to be %v, got %v", mockRuleset, resultDR.Ruleset)
			}
			if resultDR.Frames != nil {
				t.Errorf("Expected Frames to be nil, got %v", resultDR.Frames)
			}
		})

		t.Run("should handle undefined context (same as null)", func(t *testing.T) {
			detachedRuleset := NewDetachedRuleset(mockRuleset, nil)
			result := detachedRuleset.Eval(nil)

			resultDR, ok := result.(*DetachedRuleset)
			if !ok {
				t.Fatalf("Expected result to be of type *DetachedRuleset, got %T", result)
			}
			if resultDR.Ruleset != mockRuleset {
				t.Errorf("Expected Ruleset to be %v, got %v", mockRuleset, resultDR.Ruleset)
			}
			if resultDR.Frames != nil {
				t.Errorf("Expected Frames to be nil, got %v", resultDR.Frames)
			}
		})

		t.Run("should handle context with undefined frames (nil map)", func(t *testing.T) {
			detachedRuleset := NewDetachedRuleset(mockRuleset, nil)
			contextWithUndefinedFrames := map[string]interface{}{} // No frames key
			result := detachedRuleset.Eval(contextWithUndefinedFrames)

			resultDR, ok := result.(*DetachedRuleset)
			if !ok {
				t.Fatalf("Expected result to be of type *DetachedRuleset, got %T", result)
			}
			if resultDR.Ruleset != mockRuleset {
				t.Errorf("Expected Ruleset to be %v, got %v", mockRuleset, resultDR.Ruleset)
			}
			if resultDR.Frames != nil {
				t.Errorf("Expected Frames to be nil, got %v", resultDR.Frames)
			}
		})

		t.Run("should handle empty frames array in context", func(t *testing.T) {
			detachedRuleset := NewDetachedRuleset(mockRuleset, nil)
			contextWithEmptyFrames := map[string]interface{}{"frames": []interface{}{}}
			result := detachedRuleset.Eval(contextWithEmptyFrames)

			resultDR, ok := result.(*DetachedRuleset)
			if !ok {
				t.Fatalf("Expected result to be of type *DetachedRuleset, got %T", result)
			}
			if resultDR.Ruleset != mockRuleset {
				t.Errorf("Expected Ruleset to be %v, got %v", mockRuleset, resultDR.Ruleset)
			}
			if len(resultDR.Frames) != 0 {
				t.Errorf("Expected Frames to be empty, got %v", resultDR.Frames)
			}
		})
	})

	t.Run("callEval", func(t *testing.T) {
		t.Run("should evaluate the ruleset with concatenated frames when frames exist", func(t *testing.T) {
			frames := []interface{}{"frame1", "frame2"}
			detachedRuleset := NewDetachedRuleset(mockRuleset, frames)
			result := detachedRuleset.CallEval(mockContext)
			// The mock Eval returns itself by default
			if result != mockRuleset {
				t.Errorf("Expected result to be mockRuleset (%v), got %v", mockRuleset, result)
			}
		})

		t.Run("should evaluate the ruleset with original context when no frames exist", func(t *testing.T) {
			detachedRuleset := NewDetachedRuleset(mockRuleset, nil)
			result := detachedRuleset.CallEval(mockContext)
			if result != mockRuleset {
				t.Errorf("Expected result to be mockRuleset (%v), got %v", mockRuleset, result)
			}
		})

		t.Run("should handle undefined context (nil)", func(t *testing.T) {
			detachedRuleset := NewDetachedRuleset(mockRuleset, nil)
			result := detachedRuleset.CallEval(nil)
			if result != mockRuleset {
				t.Errorf("Expected result to be mockRuleset (%v), got %v", mockRuleset, result)
			}
		})

		t.Run("should handle null context", func(t *testing.T) {
			detachedRuleset := NewDetachedRuleset(mockRuleset, nil)
			result := detachedRuleset.CallEval(nil)
			if result != mockRuleset {
				t.Errorf("Expected result to be mockRuleset (%v), got %v", mockRuleset, result)
			}
		})

		t.Run("should handle ruleset eval returning null", func(t *testing.T) {
			// Use the mock's ability to customize Eval return value
			nullReturningRuleset := newMockEvaluableNode()
			nullReturningRuleset.EvalFunc = func(ctx interface{}) interface{} {
				return nil
			}
			detachedRuleset := NewDetachedRuleset(nullReturningRuleset, nil)
			result := detachedRuleset.CallEval(mockContext)
			if result != nil {
				t.Errorf("Expected result to be nil, got %v", result)
			}
		})

		t.Run("should handle ruleset eval returning undefined (same as null)", func(t *testing.T) {
			undefinedReturningRuleset := newMockEvaluableNode()
			undefinedReturningRuleset.EvalFunc = func(ctx interface{}) interface{} {
				return nil
			}
			detachedRuleset := NewDetachedRuleset(undefinedReturningRuleset, nil)
			result := detachedRuleset.CallEval(mockContext)
			if result != nil {
				t.Errorf("Expected result to be nil, got %v", result)
			}
		})

		t.Run("should handle context with undefined frames (nil map)", func(t *testing.T) {
			detachedRuleset := NewDetachedRuleset(mockRuleset, nil)
			contextWithUndefinedFrames := map[string]interface{}{}
			result := detachedRuleset.CallEval(contextWithUndefinedFrames)
			if result != mockRuleset {
				t.Errorf("Expected result to be mockRuleset (%v), got %v", mockRuleset, result)
			}
		})

		t.Run("should handle context with null frames", func(t *testing.T) {
			detachedRuleset := NewDetachedRuleset(mockRuleset, nil)
			contextWithNullFrames := map[string]interface{}{"frames": nil}
			result := detachedRuleset.CallEval(contextWithNullFrames)
			if result != mockRuleset {
				t.Errorf("Expected result to be mockRuleset (%v), got %v", mockRuleset, result)
			}
		})

		t.Run("should handle ruleset eval throwing an error", func(t *testing.T) {
			errorRuleset := newMockEvaluableNode()
			errorRuleset.EvalFunc = func(ctx interface{}) interface{} {
				panic("Eval failed")
			}
			detachedRuleset := NewDetachedRuleset(errorRuleset, nil)

			defer func() {
				if r := recover(); r == nil {
					t.Error("Expected panic with 'Eval failed', but got none")
				} else if r != "Eval failed" {
					t.Errorf("Expected panic with 'Eval failed', but got %v", r)
				}
			}()

			detachedRuleset.CallEval(mockContext)
		})

		t.Run("should handle ruleset eval returning a non-null value", func(t *testing.T) {
			valueRuleset := newMockEvaluableNode()
			valueRuleset.EvalFunc = func(ctx interface{}) interface{} {
				return "test value"
			}
			detachedRuleset := NewDetachedRuleset(valueRuleset, nil)

			result := detachedRuleset.CallEval(mockContext)
			if result != "test value" {
				t.Errorf("Expected result to be 'test value', got %v", result)
			}
		})

		t.Run("should handle frames.concat failure", func(t *testing.T) {
			// Create a custom type that will fail on append
			type failingFrames struct {
				items []interface{}
			}
			ff := &failingFrames{items: []interface{}{"frame1"}}
			
			detachedRuleset := NewDetachedRuleset(mockRuleset, ff.items)
			contextWithInvalidFrames := map[string]interface{}{
				"frames": make(chan int), // This will cause a type assertion failure
			}

			result := detachedRuleset.CallEval(contextWithInvalidFrames)
			if result != mockRuleset {
				t.Errorf("Expected result to be mockRuleset (%v), got %v", mockRuleset, result)
			}
		})

		t.Run("should handle context type assertion failure", func(t *testing.T) {
			detachedRuleset := NewDetachedRuleset(mockRuleset, nil)
			invalidContext := "not a map"

			result := detachedRuleset.CallEval(invalidContext)
			if result != mockRuleset {
				t.Errorf("Expected result to be mockRuleset (%v), got %v", mockRuleset, result)
			}
		})

		t.Run("should handle frames type assertion failure", func(t *testing.T) {
			detachedRuleset := NewDetachedRuleset(mockRuleset, nil)
			contextWithInvalidFrames := map[string]interface{}{
				"frames": "not a slice",
			}

			result := detachedRuleset.CallEval(contextWithInvalidFrames)
			if result != mockRuleset {
				t.Errorf("Expected result to be mockRuleset (%v), got %v", mockRuleset, result)
			}
		})
	})

	t.Run("accept", func(t *testing.T) {
		t.Run("should visit the ruleset with the provided visitor", func(t *testing.T) {
			detachedRuleset := NewDetachedRuleset(mockRuleset, nil)
			visited := false
			visitor := &mockVisitor{
				visit: func(node interface{}) interface{} {
					visited = true
					return node
				},
			}
			detachedRuleset.Accept(visitor)
			if !visited {
				t.Error("Expected visitor.visit to be called, but it wasn't")
			}
		})

		t.Run("should handle null visitor", func(t *testing.T) {
			defer func() {
				if r := recover(); r != nil {
					t.Errorf("The code panicked when calling Accept with nil visitor: %v", r)
				}
			}()
			detachedRuleset := NewDetachedRuleset(mockRuleset, nil)
			detachedRuleset.Accept(nil)
		})

		t.Run("should handle undefined visitor (same as null)", func(t *testing.T) {
			defer func() {
				if r := recover(); r != nil {
					t.Errorf("The code panicked when calling Accept with nil visitor: %v", r)
				}
			}()
			detachedRuleset := NewDetachedRuleset(mockRuleset, nil)
			detachedRuleset.Accept(nil)
		})

		t.Run("should handle visitor returning a different ruleset", func(t *testing.T) {
			detachedRuleset := NewDetachedRuleset(mockRuleset, nil)
			newRuleset := newMockEvaluableNode() // Use mock type here too if needed
			visitor := &mockVisitor{
				visit: func(node interface{}) interface{} {
					return newRuleset
				},
			}
			detachedRuleset.Accept(visitor)
			if detachedRuleset.Ruleset != newRuleset {
				t.Errorf("Expected Ruleset to be updated to %v, got %v", newRuleset, detachedRuleset.Ruleset)
			}
		})
	})

	t.Run("inheritance", func(t *testing.T) {
		t.Run("should have embedded Node", func(t *testing.T) {
			detachedRuleset := NewDetachedRuleset(mockRuleset, nil)
			if detachedRuleset.Node == nil {
				t.Error("Expected embedded Node field not to be nil")
			}
			if _, ok := interface{}(detachedRuleset.Node).(*Node); !ok {
				t.Errorf("Expected embedded field to be of type *Node, got %T", detachedRuleset.Node)
			}
		})

		t.Run("should have the correct type property", func(t *testing.T) {
			detachedRuleset := NewDetachedRuleset(mockRuleset, nil)
			expectedType := "DetachedRuleset"
			if detachedRuleset.Type != expectedType {
				t.Errorf("Expected Type property to be %q, got %q", expectedType, detachedRuleset.Type)
			}
		})

		t.Run("should have evalFirst set to true", func(t *testing.T) {
			detachedRuleset := NewDetachedRuleset(mockRuleset, nil)
			if !detachedRuleset.EvalFirst {
				t.Error("Expected EvalFirst property to be true, got false")
			}
		})
	})
}

// mockVisitor implements the Visitor interface for testing
type mockVisitor struct {
	visit func(interface{}) interface{}
}

func (v *mockVisitor) Visit(node interface{}) interface{} {
	if v.visit == nil {
		return node
	}
	return v.visit(node)
} 